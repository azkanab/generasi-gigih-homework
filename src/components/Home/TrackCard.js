import '../../styles/Home/TrackCard.css'

export default function TrackCard({ data }) {
    return (
        <div className="wrapper">
            <img src={data.imgUrl} alt={data.albumName + ' Album'} className="albumImage" />
		    <p className="titleText">{data.trackTitle}</p>
		    <p className="artistText">{data.artistName}</p>
		    <p className="albumText">{data.albumName}</p>
		    <button>Select</button>
		</div>
    )
}