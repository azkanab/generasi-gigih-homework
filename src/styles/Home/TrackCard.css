.wrapper {
	height: fit-content;
	width: auto;
	background-color: #F8EAD9;
	border: 1.5px solid black;
	padding: 10px;
}

.albumImage {
	height: 200px;
	width: auto;
	margin-bottom: 10px;
}

p {
	margin-bottom: 5px;	
}

.title {
	font-weight: bold;
	font-size: 20px
}

.wrapper button {
	padding: 2px 10px;
	margin-top: 10px
}