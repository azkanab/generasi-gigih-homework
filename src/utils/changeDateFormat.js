export function changeDateFormat(date) {
    var monthNames = [ "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December" ];

    var newDate = new Date(date);
    var formattedDate = `${newDate.getDate()} ${monthNames[newDate.getMonth()]} ${newDate.getFullYear()}`
    return formattedDate
}