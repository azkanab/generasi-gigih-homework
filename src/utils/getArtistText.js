export const getArtistText = (artists) => {
    let result = ''
    artists.forEach((artist, idx) => {
        if (idx === artists.length-1) {
            result += artist
        } else {
            result += `${artist}, `
        }
    })
    return result
}