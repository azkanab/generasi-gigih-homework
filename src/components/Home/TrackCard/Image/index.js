import '../../../../styles/Home/TrackCard.css'

export default function Image ({ imgUrl, imgAlt }) {
    return (
        <img src={imgUrl} alt={imgAlt} className="albumImage" />
    )
}