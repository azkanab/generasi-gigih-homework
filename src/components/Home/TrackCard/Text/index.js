import '../../../../styles/Home/TrackCard.css'

export default function Text ({ type, text }) {

    const isBoldText = () => {
        return type === 'title'
    }

    return (
        <p className={isBoldText() ? 'title' : ''}>{text}</p>
    )
}