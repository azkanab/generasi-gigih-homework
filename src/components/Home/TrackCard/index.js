import { useEffect, useState } from 'react'
// import { useRecoilState } from 'recoil'
import '../../../styles/Home/TrackCard.css'
import Image from '../../common/Image'
import Button from '../../common/Button'
import Text from '../../common/Text'
import isObjectEmpty from '../../../utils/isObjectEmpty'
// import { selectedTrackState } from '../../../state/selectedTrack'

const TRACK_CARD_TYPE = {
	unselected: "normal",
	selected: "selected"
}

const SELECT_BUTTON = ""
const IMG_PLAY_BUTTON = "/playbutton.png"
const IMG_DESELECT_BUTTON = "/remove.png"
const IMG_PLAY_ALT = "Select Button"
const IMG_DESELECT_ALT = "Deselect Button"
const IMG_PLAY_CLASS = "play-button"

export default function TrackCard({ type, data, handleClick }) {

	const [track, setTrack] = useState({})
	// const [selectedTrack, setSelectedTrack] = useRecoilState(selectedTrackState)

	const playButton = {
		imgUrl: IMG_PLAY_BUTTON,
		imgAlt: IMG_PLAY_ALT,
		imgClass: IMG_PLAY_CLASS
	}
	const deselectButton = {
		imgUrl: IMG_DESELECT_BUTTON,
		imgAlt: IMG_DESELECT_ALT,
		imgClass: IMG_PLAY_CLASS
	}

	const isSelected = () => {
		return type === TRACK_CARD_TYPE.selected
	}

	// const checkItemToDeselect = (item) => {
	// 	return item.uri !== track.uri
	// }

	const handleSelectButtonClick = () => {
		// window.location.href = track.url
		if (!isSelected()) {
			// Select Item
			// let newSelectedTrack = [...selectedTrack]
			// newSelectedTrack.push(data)
			// setSelectedTrack(newSelectedTrack)
			handleClick(data)
		} else {
			// Deselect Item
			// let newSelectedTrack = selectedTrack.filter(checkItemToDeselect)
			// setSelectedTrack(newSelectedTrack)
		}
	}

	const renderArtists = (artists) => {
		return (
			artists.map((artist, idx) => (
				idx === artists.length-1 ?
					<Text key={idx} textClass="artistText" text={artist} />
				:
					<Text key={idx} textClass="artistText" text={artist + ', '} />
			))
		)
	}

	useEffect(() => {
		if (!isObjectEmpty(data)) {
			setTrack({
				altAlbum: data.albumName + ' Album',
				imgUrl: data.imgUrl,
				title: data.trackTitle,
				artists: data.artistName,
				album: data.albumName,
				url: data.spotifyUrl,
				uri: data.uri
			})
		}
	}, [data])

    return (
		!isObjectEmpty(track) &&
		<div className="wrapper">
			<div className="image-wrapper">
				<Image imgUrl={track.imgUrl} imgAlt={track.altAlbum} imgClass="albumImage" />
				<div className="button-wrapper">
					{!isSelected() ?
						<Button primary img={playButton} text={SELECT_BUTTON} handleClick={handleSelectButtonClick} />
					:
						<Button secondary img={deselectButton} text={SELECT_BUTTON} handleClick={handleSelectButtonClick} />
					}
				</div>
			</div>
			<div className="detail-container">
				<div className="text-wrapper">
					<Text textClass="title" text={track.title} />
					{renderArtists(track.artists)}
					<Text textClass="albumText" text={track.album} />
				</div>
			</div>
		</div>
		
    )
}