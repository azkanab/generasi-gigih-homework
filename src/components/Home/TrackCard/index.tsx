import { useEffect, useState } from 'react'
import '../../../styles/Home/TrackCard.css'
import Image from '../../common/Image'
import Button from '../../common/Button'
import Text from '../../common/Text'
import isObjectEmpty from '../../../utils/isObjectEmpty'
import { getArtistText } from '../../../utils/getArtistText'

const TRACK_CARD_TYPE = {
	unselected: "normal",
	selected: "selected"
}

const SELECT_BUTTON = ""
const IMG_PLAY_BUTTON = "/playbutton.png"
const IMG_DESELECT_BUTTON = "/remove.png"
const IMG_PLAY_ALT = "Select Button"
const IMG_DESELECT_ALT = "Deselect Button"
const IMG_PLAY_CLASS = "play-button"

interface Data {
	id: string,
	uri: string,
	imgUrl: string,
	trackTitle: string,
	artistName: string[],
	albumName: string,
	spotifyUrl: string
}

interface Props {
	type: string,
	data: Data,
	handleClick: (data: Data) => void
}

interface TrackState {
	altAlbum?: string,
	imgUrl?: string,
	title?: string,
	artists?: string[],
	album?: string,
	url?: string,
	uri?: string
}

export default function TrackCard({ type, data, handleClick }: Props) {

	const [track, setTrack] = useState<TrackState>({})

	const playButton = {
		imgUrl: IMG_PLAY_BUTTON,
		imgAlt: IMG_PLAY_ALT,
		imgClass: IMG_PLAY_CLASS
	}
	const deselectButton = {
		imgUrl: IMG_DESELECT_BUTTON,
		imgAlt: IMG_DESELECT_ALT,
		imgClass: IMG_PLAY_CLASS
	}

	const isSelected = () => {
		return type === TRACK_CARD_TYPE.selected
	}

	const handleSelectButtonClick = () => {
		if (!isSelected()) {
			handleClick(data)
		}
	}

	useEffect(() => {
		if (!isObjectEmpty(data)) {
			setTrack({
				altAlbum: data.albumName + ' Album',
				imgUrl: data.imgUrl,
				title: data.trackTitle,
				artists: data.artistName,
				album: data.albumName,
				url: data.spotifyUrl,
				uri: data.uri
			})
		}
	}, [data])

    return (
		!isObjectEmpty(track) ?
		<div id="track-card" className="track-card">
			<div className="track-card__image-wrapper">
				<Image imgUrl={track.imgUrl} imgAlt={track.altAlbum} imgClass="track-card__album-image" />
				<div className="track-card__button-wrapper">
					{!isSelected() ?
						<Button primary img={playButton} text={SELECT_BUTTON} handleClick={handleSelectButtonClick} />
					:
						<Button secondary img={deselectButton} text={SELECT_BUTTON} handleClick={handleSelectButtonClick} />
					}
				</div>
			</div>
			<div className="track-card__detail">
				<div className="track-card__text-wrapper">
					<Text textClass="card__title" text={track.title} />
					<Text textClass="card__artist" text={getArtistText(track.artists)} />
					<Text textClass="card__album" text={track.album} />
				</div>
			</div>
		</div>
		: null
    )
}