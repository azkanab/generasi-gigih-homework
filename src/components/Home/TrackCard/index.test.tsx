import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import TrackCard from "./index";
import { getArtistText } from "../../../utils/getArtistText";

const type = "normal"

const data = {
    id: '123',
    uri: 'spotify:123',
    imgUrl: 'https://www.rspca.org.uk/documents/1494939/0/what+to+do+with+stray+cats+and+kittens+%283%29.jpg/886bc0d5-1dc5-d2bf-eabd-473fd4d99886?t=1618404272031',
    trackTitle: 'Catty Cat',
    artistName: ['Adylan Roaffa', 'Azka Mumtaz'],
    albumName: 'Lovey Dovey',
    spotifyUrl: 'https://www.instagram.com/azkanab/'
}

const onClick = jest.fn()
const handleSelectButtonClick = jest.fn()

const IMG_PLAY_BUTTON = "/playbutton.png"
const IMG_PLAY_ALT = "Select Button"
const IMG_PLAY_CLASS = "play-button"

const playButton = {
    imgUrl: IMG_PLAY_BUTTON,
    imgAlt: IMG_PLAY_ALT,
    imgClass: IMG_PLAY_CLASS
}

beforeEach(() => {
    render (
        <TrackCard type={type} data={data} handleClick={onClick} />
    )
})

test('Check if Track Card is rendered', () => {
    const { container } = render (
        <TrackCard type={type} data={data} handleClick={onClick} />
    )
    expect(container.firstChild).toMatchSnapshot(`
        <div id="track-card" class="track-card">
            <div class="track-card__image-wrapper">
                <div class="image">
                    <div class="image__placeholder-container">
                        <img src="/unknown_playlist.png" alt="loading" class="image__placeholder" />
                    </div>
                    <img src="${data.imgUrl}" alt="${data.albumName + ' Album'}" class="track-card__album-image image__display-none" />
                </div>
                <div class="track-card__button-wrapper">
                    <button onClick=${handleSelectButtonClick} class="button-primary">
                        <img src="${playButton.imgUrl}" alt="${playButton.imgAlt}" class="${playButton.imgClass}" />}
                    </button>
                </div>
            </div>
            <div class="track-card__detail">
                <div class="track-card__text-wrapper">
                    <p class="card__title">${data.trackTitle}</p>
                    <p class="card__artist">${getArtistText(data.artistName)}</p>
                    <p class="card__album"/>${data.albumName}</p>
                </div>
            </div>
        </div>
    `)
})

test('Check if Track Card renders image', () => {
    const imgElement = screen.getByAltText(data.albumName + ' Album')
    expect(imgElement).toBeInTheDocument()
    expect(imgElement.getAttribute('src')).toBe(data.imgUrl)
})

test('Check if Track Card renders the title', () => {
    const titleElement = screen.getByText(data.trackTitle)
    expect(titleElement).toBeInTheDocument()
})

test('Check if Track Card renders the artist', () => {
    const artistName = screen.getByText(getArtistText(data.artistName))
    expect(artistName).toBeInTheDocument()
})

test('Check if Track Card renders the album name', () => {
    const albumName = screen.getByText(data.albumName)
    expect(albumName).toBeInTheDocument()
})

test('Check if Track Card renders playButton', () => {
    const buttonPlayElement = screen.getByRole('button')
    expect(buttonPlayElement).toBeInTheDocument()
    expect(buttonPlayElement.getAttribute('class')).toBe('button-primary')
})

test('Check if onClick function is called when user click the playButton', () => {
    const buttonPlayElement = screen.getByRole('button')

    const leftClick = { button: 0 }
    userEvent.click(buttonPlayElement, leftClick)

    expect(onClick).toBeCalledTimes(1)
})