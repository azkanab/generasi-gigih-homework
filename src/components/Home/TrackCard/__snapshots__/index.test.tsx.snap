// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if Track Card is rendered: 
        <div id="track-card" class="wrapper">
            <div class="image-wrapper">
                <div class="image">
                    <div class="image-placeholder-container">
                        <img src="/unknown_playlist.png" alt="loading" class="image-placeholder" />
                    </div>
                    <img src="https://www.rspca.org.uk/documents/1494939/0/what+to+do+with+stray+cats+and+kittens+%283%29.jpg/886bc0d5-1dc5-d2bf-eabd-473fd4d99886?t=1618404272031" alt="Lovey Dovey Album" class="image-album display-none" />
                </div>
                <div class="button-wrapper">
                    <button onClick=function () {
        return fn.apply(this, arguments);
      } class="button-primary">
                        <img src="/playbutton.png" alt="Select Button" class="play-button" />}
                    </button>
                </div>
            </div>
            <div class="detail-container">
                <div class="text-wrapper">
                    <p class="title">Catty Cat</p>
                    <p class="artistText">Adylan Roaffa, Azka Mumtaz</p>
                    <p class="albumText"/>Lovey Dovey</p>
                </div>
            </div>
        </div>
     1`] = `
<div
  class="wrapper"
  id="track-card"
>
  <div
    class="image-wrapper"
  >
    <div
      class="image"
    >
      <div
        class="image-placeholder-container"
      >
        <img
          alt="loading"
          class="image-placeholder"
          src="/unknown_playlist.png"
        />
      </div>
      <img
        alt="Lovey Dovey Album"
        class="image-album image-display-none"
        src="https://www.rspca.org.uk/documents/1494939/0/what+to+do+with+stray+cats+and+kittens+%283%29.jpg/886bc0d5-1dc5-d2bf-eabd-473fd4d99886?t=1618404272031"
      />
    </div>
    <div
      class="button-wrapper"
    >
      <button
        class="button-primary"
      >
        
        <img
          alt="Select Button"
          class="play-button"
          src="/playbutton.png"
        />
      </button>
    </div>
  </div>
  <div
    class="detail-container"
  >
    <div
      class="text-wrapper"
    >
      <p
        class="title"
      >
        Catty Cat
      </p>
      <p
        class="artistText"
      >
        Adylan Roaffa, Azka Mumtaz
      </p>
      <p
        class="albumText"
      >
        Lovey Dovey
      </p>
    </div>
  </div>
</div>
`;
